### PR TITLE
Clarify the change of enc -S behavior in 3.0

### LINKED ISSUE #19730: openssl enc -S behavior changed in 3.0 producing files not readable with 1.1.1
Here is a way to demonstrate and reproduce the issue.

First verify openssl version and prepare the environment on the host:
```
host> openssl version
OpenSSL 3.0.7 1 Nov 2022 (Library: OpenSSL 3.0.7 1 Nov 2022)

host> mkdir /tmp/openssl-test

host> echo 'Just say narf.
Just say narf!
Just say NARF!' > /tmp/openssl-test/test-input.txt

host> cat /tmp/openssl-test/test-input.txt
Just say narf.
Just say narf!
Just say NARF!

host> docker run --name ubuntu2 -d -v /tmp/openssl-test:/mnt -it ubuntu:18.04
host> docker exec -it ubuntu2 bash
```

Check openssl version in the container and encrypt the files (also verify that decryption works):
```
container> apt update && apt install openssl
container> openssl version
OpenSSL 1.1.1  11 Sep 2018

container> ls -alFh /mnt/
total 8.0K
drwxr-xr-x 2 1000 1000   60 Nov 22 12:45 ./
drwxr-xr-x 1 root root 4.0K Nov 22 12:43 ../
-rw-r--r-- 1 1000 1000   45 Nov 22 12:46 test-input.txt

container> cat /mnt/test-input.txt
Just say narf.
Just say narf!
Just say NARF!

container> openssl enc -p -aes256 -a -S CEED -pbkdf2 -iter 20000 -pass pass:NARF -in /mnt/test-input.txt -out /mnt/test-output.enc
hex string is too short, padding with zero bytes to length
salt=CEED000000000000
key=F3F404D5ABC2D888ADD1133ED9B3CCA77E4317369462D16FE70539643184E21D
iv =D00113672386E8CE59518BECD6835719

container> cat /mnt/test-output.enc
U2FsdGVkX1/O7QAAAAAAAMafQdZAkgr0JVD+lONLKNvB55ZaEaJz1hQUKIt2tIom
1XhCTf4axktE6CkX+j9mDQ==

container> openssl enc -d -p -aes256 -a -S CEED -pbkdf2 -iter 20000 -pass pass:NARF -in /mnt/test-output.enc -out /mnt/test-decrypt.txt.v1
salt=CEED000000000000
key=F3F404D5ABC2D888ADD1133ED9B3CCA77E4317369462D16FE70539643184E21D
iv =D00113672386E8CE59518BECD6835719

container> cat /mnt/test-decrypt.txt.v1
Just say narf.
Just say narf!
Just say NARF!
``` 

Decrypt the file on the host:
``` 
host> openssl enc -d -p -aes256 -a -S CEED -pbkdf2 -iter 20000 -pass pass:NARF -in /tmp/openssl-test/test-output.enc -out /tmp/openssl-test/test-decrypt.txt.v3
hex string is too short, padding with zero bytes to length
salt=CEED000000000000
key=F3F404D5ABC2D888ADD1133ED9B3CCA77E4317369462D16FE70539643184E21D
iv =D00113672386E8CE59518BECD6835719

host> cat /tmp/openssl-test/test-decrypt.txt.v3
KG^:���7▒��=AmV��
                 gf�����Ʞ�]Sust say narf!
Just say NARF!
``` 
-> **The first few bytes of the output are corrupted / garbage**

Try again with openssl 3 only:
``` 
host> openssl enc -p -aes256 -a -S CEED -pbkdf2 -iter 20000 -pass pass:NARF -in /tmp/openssl-test/test-input.txt -out /tmp/openssl-test/test-output.enc.v3
hex string is too short, padding with zero bytes to length
salt=CEED000000000000
key=F3F404D5ABC2D888ADD1133ED9B3CCA77E4317369462D16FE70539643184E21D
iv =D00113672386E8CE59518BECD6835719

host> cat /tmp/openssl-test/test-output.enc.v3
xp9B1kCSCvQlUP6U40so28HnlloRonPWFBQoi3a0iibVeEJN/hrGS0ToKRf6P2YN

host> openssl enc -d -p -aes256 -a -S CEED -pbkdf2 -iter 20000 -pass pass:NARF -in  /tmp/openssl-test/test-output.enc.v3 -out /tmp/openssl-test/test-v3-decrypt.txt
hex string is too short, padding with zero bytes to length
salt=CEED000000000000
key=F3F404D5ABC2D888ADD1133ED9B3CCA77E4317369462D16FE70539643184E21D
iv =D00113672386E8CE59518BECD6835719

host> cat /tmp/openssl-test/test-v3-decrypt.txt
Just say narf.
Just say narf!
Just say NARF!
``` 

_Host OS_: Arch Linux
_Host openssl version_: OpenSSL 3.0.7 1 Nov 2022 (Library: OpenSSL 3.0.7 1 Nov 2022)
_Container OS_: Ubuntu 18.04
_Container openssl version_: OpenSSL 1.1.1  11 Sep 2018

### PR DESCRIPTION
**Pull request by t8m**
Fixes #19730

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

File: doc/man1/openssl-enc.pod.in [hunk 1]
```diff
@@ -451,6 +451,13 @@ The B<-list> option was added in OpenSSL 1.1.1e.
 
 The B<-ciphers> and B<-engine> options were deprecated in OpenSSL 3.0.
 
+Since OpenSSL 3.0 using the B<-S> option makes the encrypted file to not
+include the salt into the encrypted file and during the decryption the
+salt is no longer expected to be present in the file. To decrypt files
+encrypted on OpenSSL 1.1.1 with B<-S>, just do not use the B<-S> option.
+To produce files that can be decrypted on OpenSSL 1.1.1 do not use the
+B<-S> option, the salt will be generated randomly.
+
```
Comment (paulidale) [inline doc/man1/openssl-enc.pod.in:460]: What about:

> OpenSSL 3.0 changed the way the B<-S> operates to not include the specified salt in the encrypted file.  This prevents decryption using OpenSSL 1.1.1.  To produce files that can be decrypted using OpenSSL 1.1.1 do not use the
B<-S> option with OpenSSL 3.0.  In this case the salt will be generated randomly.

I think this should be in the NOTES section not the HISTORY.

**Review by paulidale — CHANGES_REQUESTED** [review]

File: doc/man1/openssl-enc.pod.in [hunk 2]
```diff
@@ -451,6 +451,13 @@ The B<-list> option was added in OpenSSL 1.1.1e.
 
 The B<-ciphers> and B<-engine> options were deprecated in OpenSSL 3.0.
 
+Since OpenSSL 3.0 using the B<-S> option makes the encrypted file to not
```
Comment (slontis) [inline doc/man1/openssl-enc.pod.in:454]: delete 'to'

**Comment by t8m** [comment]
@slontis @paulidale better now?

File: doc/man1/openssl-enc.pod.in (hunk 1 above)
Comment (t8m) [reply doc/man1/openssl-enc.pod.in]: I've partially used your proposal also moved to NOTES.

File: doc/man1/openssl-enc.pod.in (hunk 2 above)
Comment (t8m) [reply doc/man1/openssl-enc.pod.in]: Reworded according to @paulidale's proposal.

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
@slontis?  Review?

File: doc/man1/openssl-enc.pod.in [hunk 3]
```diff
@@ -256,6 +256,15 @@ All RC2 ciphers have the same key and effective key length.
 
 Blowfish and RC5 algorithms use a 128 bit key.
 
+OpenSSL 3.0 changed the way the B<-S> option operates to not include the
```
Comment (slontis) [inline doc/man1/openssl-enc.pod.in:259]: This seems a bit convoluted.. What about this?

For encryption purposes specifying the -salt option without the -S option will write the randomly generated salt to the encrypted file. If you use -S then the supplied salt is not written to the encrypted file. Note that prior to OpenSSL 3.0 the -S option would write to the encrypted file.

For decryption purposes if the salt used for encryption is not present in the encrypted file then it must be supplied using the -S option.

Comment (vdukhovni) [reply doc/man1/openssl-enc.pod.in:259]: Sadly, I don't find the above to be an improvement. But I think progress is possible.

```
Please note that OpenSSL 3.0 changed the effect of the B<-S> option.
Any explicit salt value specified via this option is no longer prepended to the
ciphertext when encrypting, and must again be explicitly provided when decrypting.
Conversely, when the B<-S> option is used during decryption, the ciphertext
is expected to not have a prepended salt value.

When using OpenSSL 3.0 or later to decrypt data that was encrypted with an
explicit salt under OpenSSL 1.1.1 do not the B<-S> option, the salt will then be
read from the ciphertext.
To generate ciphertext that can be decrypted with OpenSSL 1.1.1 do not use
the B<-S> option, the salt will be then be generated randomly and prepended
to the output.
```

**Comment by t8m** [comment]
I've used the text as suggested by @vdukhovni - I've made you the commit author as it is your suggestion verbatim.

File: doc/man1/openssl-enc.pod.in (hunk 3 above)
Comment (vdukhovni) [reply doc/man1/openssl-enc.pod.in]: I hope you added the missing "use" in the 2nd line of the 2nd paragraph: "... OpenSSL 1.1.1 do not [use] the B<-S>..."

File: doc/man1/openssl-enc.pod.in [hunk 4]
```diff
@@ -256,6 +256,19 @@ All RC2 ciphers have the same key and effective key length.
...
 Blowfish and RC5 algorithms use a 128 bit key.
 
+Please note that OpenSSL 3.0 changed the effect of the B<-S> option.
+Any explicit salt value specified via this option is no longer prepended to the
+ciphertext when encrypting, and must again be explicitly provided when decrypting.
+Conversely, when the B<-S> option is used during decryption, the ciphertext
+is expected to not have a prepended salt value.
+
+When using OpenSSL 3.0 or later to decrypt data that was encrypted with an
+explicit salt under OpenSSL 1.1.1 do not the B<-S> option, the salt will then be
```
Comment (vdukhovni) [inline doc/man1/openssl-enc.pod.in:266]:
```suggestion
explicit salt under OpenSSL 1.1.1 do not use the B<-S> option, the salt will then be
```

Comment (t8m) [reply doc/man1/openssl-enc.pod.in]: Done

**Review by slontis — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
This was already merged. Closing.

**Pull request closed without merging**